File: architecture/data/src/main/java/com/thepacific/data/http/Source.java
package com.thepacific.data.http;

public final class Source<T> {

  public final Status status;
  public final Throwable error;
  public final T data;

  private Source(Status status, Throwable error, T data) {
    this.status = status;
    this.error = error;
    this.data = data;
  }

  public static <T> Source<T> inProgress() {
    return new Source<>(Status.IN_PROGRESS, null, null);
  }

  public static <T> Source<T> success(T data) {
    return new Source<>(Status.SUCCESS, null, data);
  }

  public static <T> Source<T> failure(Throwable error) {
    return new Source<>(Status.ERROR, error, null);
  }

  public static <T> Source<T> irrelevant() {
    return new Source<>(Status.IRRELEVANT, null, null);
  }
}

File: architecture/arch-kotlin/src/main/java/com/pacific/arch/kotlin/data/Source.kt
package com.pacific.arch.kotlin.data

data class Source<out T> private constructor(val status: Status,
                                             val error: Throwable?,
                                             val data: T?) {
    companion object {
        @JvmStatic
        fun <T> inProgress(): Source<T> {
            return Source<T>(Status.IN_PROGRESS, null, null)
        }

        @JvmStatic
        fun <T> success(data: T): Source<T> {
            return Source(Status.SUCCESS, null, data)
        }

        @JvmStatic
        fun <T> failure(error: Throwable): Source<T> {
            return Source<T>(Status.ERROR, error, null)
        }

        @JvmStatic
        fun <T> irrelevant(): Source<T> {
            return Source<T>(Status.IRRELEVANT, null, null)
        }
    }
}